### PR TITLE
Introduces on_request and on_complete methods in Faraday::Middleware.

### PR DESCRIPTION
**Pull request by iMacTia**
## Description
* Introduces `on_request` and `on_complete` methods in `Faraday::Middleware`
* Adds documentation on how to leverage the new methods

## Additional Notes
Unit tests for `on_complete` were a bit complicated to write as they rely on the callback system of the `Rack::Builder`.
It would be better to write integration tests for that, but given the limited amount of time I couldn't make it today.

### CODE REVIEW

File: docs/middleware/custom.md [hunk 1]
```diff
@@ -43,3 +42,16 @@ later, response. Some keys are:
...
 ```
+
+### Faraday::Middleware
+
+There's an easier way to write middleware, and it's also the recommended one: make your middleware subclass `Faraday::Middleware`.
+`Faraday::Middleware` already implements the `call` method for you and looks for two methods in your subclass: `on_request(env) and `on_complete(env)`.
+`on_request` is called when the request is being built and is given the `env` representing the request.
+`on_complete` is called after the response has been received (that's right, it already supports parallel mode!) and receives the `env` of the response.
+
+### Do I need to override `call`?
```
Comment (olleolleolle) [inline docs/middleware/custom.md:53]: Minor: `#call`, `#on_request` etc could be names.

**Review by olleolleolle — APPROVED** [review]
🎂

File: docs/middleware/custom.md (hunk 1 above)
Comment (iMacTia) [reply docs/middleware/custom.md:53]: yeah I like that, it makes it clear that they're method unlike `env` 👍

File: docs/middleware/custom.md [hunk 2]
```diff
@@ -43,3 +42,16 @@ later, response. Some keys are:
...
+### Faraday::Middleware
+
+There's an easier way to write middleware, and it's also the recommended one: make your middleware subclass `Faraday::Middleware`.
+`Faraday::Middleware` already implements the `#call` method for you and looks for two methods in your subclass: `#on_request(env) and `#on_complete(env)`.
+`#on_request` is called when the request is being built and is given the `env` representing the request.
+`#on_complete` is called after the response has been received (that's right, it already supports parallel mode!) and receives the `env` of the response.
+
+### Do I need to override `#call`?
+
+For the majority of middleware, it's not necessary to override the `#call` method, as you can simply use `#on_request` and `#on_response`.
```
Comment (natematykiewicz) [inline docs/middleware/custom.md:55]: This on_response should be on_complete, right?

Comment (iMacTia) [reply docs/middleware/custom.md:55]: YES 🙈

Comment (iMacTia) [reply docs/middleware/custom.md:55]: Luckily, that was already spot and fixed 🙌 
https://github.com/lostisland/faraday/commit/4e9f4b06f1a1d46fac45663ab3a857aa60b79012

Comment (natematykiewicz) [reply docs/middleware/custom.md:55]: Oh nice! I was just reading through the PRs referenced in the release notes. I didn't think to diff the two versions. 😬

**Pull request merged**